aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -69,14 +132,54 @@ public void flush() throws IOException {
 
   @Override
   public void write(int b) throws IOException {
+    if (stream.getCount() >= multiPartSize) {
+      newStream();
+      uploadParts();
+    }
+
     stream.write(b);
     pos += 1;
   }
 
   @Override
   public void write(byte[] b, int off, int len) throws IOException {
-    stream.write(b, off, len);
+    int remaining = len;
+    int relativeOffset = off;
+
+    // Write the remainder of the part size to the staging file
+    // and continue to write new staging files if the write is
+    // larger than the part size.
+    while (stream.getCount() + remaining > multiPartSize) {
+      int writeSize = multiPartSize - (int) stream.getCount();
```

Comment:
maybe also try to initialize the multipart upload up here so that the individual parts can get kicked off immediately during the loop?

(That currently returns immediately if this is the first chunk right?)

Reply:
I feel like this is really unlikely to be an issue.  The loop is mostly for completeness, but for you to hit a second iteration, the write size would have to be 2x the page size and given this is internally wrapped in a buffered output stream, the part size would need to be less than 4K (less than half the output buffer size).  

I think your right that it wouldn't start uploading parts until after the loop, but I just don't feel like it's worth the extra complexity.

Reply:
Makes sense!